### PR TITLE
Support signaling core.Start execs.

### PR DESCRIPTION
**Pull request by sipsma**
This adds a new core.SendSignal task that enables arbitrary signals to
be sent to an async Start exec. It also adds related support to
core.Stop for specifying a timeout to wait for the exec to exit on its
own before sending SIGKILL.

The main goal here is to support ending Start execs gracefully instead
of being forced to always just SIGKILL them as part of core.Stop. It
should be useful in general for any other use cases that call for
signals to be sent to these execs (i.e. SIGWINCH for future tty
support).

Among other things, this is needed to support running Jaeger via
core.Start (without a graceful stop, its data gets corrupted).

Signed-off-by: Erik Sipsma <erik@sipsma.dev>

### CODE REVIEW

File: pkg/dagger.io/dagger/core/exec.cue [hunk 1]
```diff
@@ -125,3 +128,48 @@ import "dagger.io/dagger"
...
+#SendSignal: {
+	$dagger: task: _name: "SendSignal"
+
+	input: #Start
+
+	signal: or([ for name, signal in _signals {signal}])
+}
+
+_signals: {
+	SIGHUP:    1
```
Comment (TomChv) [inline pkg/dagger.io/dagger/core/exec.cue:144]: > Nice, that's much cleaner! I also had to include a line that embedded _#signals on the top level of the file so outside callers could still say things like core.#SIGHUP, but that was easy too (thanks cue!)

Currently he will do `core.SIGHUP`, is it still ok for you?

File: plan/task/exec.go [hunk 2]
```diff
@@ -150,19 +153,46 @@ func (t *stopAsyncExecTask) Run(ctx context.Context, pctx *plancontext.Context,
...
+		return nil, err
+	}
+
 	lg := log.Ctx(ctx)
 
-	exitCode, err := s.StopContainer(ctx, ctrID)
+	exitCode, err := s.StopContainer(ctx, ctrID, time.Duration(timeout))
 	if err != nil {
-		return nil, fmt.Errorf("failed to stop async exec %s: %w", ctrID, err)
+		return nil, fmt.Errorf("failed to stop exec %s: %w", ctrID, err)
```
Comment (TomChv) [inline plan/task/exec.go:165]: It is still an asynchronous execution, why did you remove that from log & error??

**Review by TomChv — COMMENTED** [review]
Left minor comments, that looks really good! I'll give it a try for my final review

File: pkg/dagger.io/dagger/core/exec.cue (hunk 1 above)
Comment (sipsma) [reply pkg/dagger.io/dagger/core/exec.cue:144]: Ah yeah, I don't think making it a definition does anything since its just a concrete number.

File: plan/task/exec.go (hunk 2 above)
Comment (sipsma) [reply plan/task/exec.go:165]: Changed it because even though it is asynchronous, that's not the terminology we ended up using, so I thought that using `async` in the user-facing error message might be slightly confusing.

**Review by TomChv — COMMENTED** [review]
I just tried `core.#Start` but it didn't run the container asynchronuously.

Maybe I have misunderstood the usage but I expect that the container and the process do not stop until I stop dagger or send a [signal SIGINT](https://github.com/docker-library/postgres/blob/1ae967e6437ff5bedd1d977a06b78a9b9fc13df8/14/alpine/Dockerfile#L179).

Here's my plan :

```cue
package main

import (
	"dagger.io/dagger"
	"dagger.io/dagger/core"
)

dagger.#Plan & {
	actions: {
		image: core.#Pull & {
			source: "postgres:14.3-alpine"
		}

		test: {
			start: core.#Start & {
				input: image.output
				args: ["docker-entrypoint.sh", "postgres"]
				env: {
					POSTGRES_PASSWORD: "password"
					POSTGRES_USER:     "foo"
					POSTGRES_DB:       "test"
					image.config.env
				}
			}
		}
	}
}
```

Here's a repro case with a docker command
```shell
docker run -p 5432:5432 -e POSTGRES_PASSWORD="password" -e POSTGRES_USER="foo" -e POSTGRES_DB="test" postgres:14.3-alpine docker-entrypoint.sh postgres
```

Do I miss anything? That would be really cool to run a side database, specially for test purpose

**Comment by sipsma** [comment]
@TomChv That's the expected behavior. Being asynchronous means that we don't wait for it to complete, the `core.#Start` task is considered completed once it has started the process and then the process continues to run in the background. In your example, you start the database but then there are no more tasks to run after it, so dagger is done and exits (and kills the container). Your `docker run` command is different because `docker run` is by default a synchronous command (like `core.#Exec`). What you are doing with `core.#Start` currently is more like if you ran `docker run ... &` and then exited your shell right away (causing the background process to be killed).

For the use case of running a database to be used by other tasks, what's expected is that you use `core.#Start` like you do but then actually run tests as separate actions. Dagger will wait for those tests to finish running (while the database runs in the background). Once the tests are done, the background database process will be killed as dagger exits.

There are test cases that verify all this behavior works as expected: https://github.com/dagger/dagger/blob/main/tests/tasks/exec/start_stop_exec.cue

File: plan/task/exec.go [hunk 3]
```diff
@@ -22,6 +24,7 @@ func init() {
 	Register("Exec", func() Task { return &execTask{} })
 	Register("Start", func() Task { return &asyncExecTask{} })
 	Register("Stop", func() Task { return &stopAsyncExecTask{} })
+	Register("SendSignal", func() Task { return &sendSignalTask{} })
```
Comment (TomChv) [inline plan/task/exec.go:27]: This file become bigger and bigger after each improvement. If we continue like this, `exec.go` will be unmaintainable.

We should maybe work on a refactor of the file tree to make it more sustainable with that kind of improvement.

However, that should be done in a further PR

File: plan/task/exec.go [hunk 4]
```diff
@@ -150,19 +153,46 @@ func (t *stopAsyncExecTask) Run(ctx context.Context, pctx *plancontext.Context,
...
+	if err != nil {
+		return nil, err
+	}
+	sig := syscall.Signal(sigVal)
+
+	if err := s.SignalContainer(ctx, ctrID, sig); err != nil {
+		return nil, fmt.Errorf("failed to send signal %d to exec %s: %w", sig, ctrID, err)
+	}
+
+	return compiler.NewValue(), nil
```
Comment (TomChv) [inline plan/task/exec.go:193]: Why not just return nil since the value is unused after that? or v?

**Review by TomChv — APPROVED** [review]
This improvement is really amazing! I think we should write a piece of documentation about it because it's quite hard to understand how use it.
Feel free to open an issue to keep track of this ;)

PR LGTM on my side, I left questions but it is mergeable IMO

**Comment by sipsma** [comment]
> This improvement is really amazing! I think we should write a piece of documentation about it because it's quite hard to understand how use it. Feel free to open an issue to keep track of this ;)
> 
> PR LGTM on my side, I left questions but it is mergeable IMO

Thanks Tom! The remaining issues, including creating all the docs, are tracked here: https://github.com/dagger/dagger/issues/2398

Overall, the goal of these tasks in `core` is to create a set of flexible primitives. Then, we can make higher level interfaces in `universe` packages that are more opinionated and easy to use.

File: plan/task/exec.go (hunk 3 above)
Comment (sipsma) [reply plan/task/exec.go:27]: Agree, I have more PRs pending this one, can update them to split things up here a bit more

File: plan/task/exec.go (hunk 4 above)
Comment (sipsma) [reply plan/task/exec.go:193]: I think I tried returning just `nil` but got a nil pointer exception when the caller of this method tried to unify it with the original value. I guess you could have `v` unify with itself too but not sure there's much difference between that and returning a new empty value here.

File: plan/task/exec.go (hunk 3 above)
Comment (TomChv) [reply plan/task/exec.go:27]: Great!

File: plan/task/exec.go (hunk 4 above)
Comment (TomChv) [reply plan/task/exec.go:193]: Okay, that's strange but it's ok to just return a new value then, thank you!

Comment (helderco) [reply plan/task/exec.go:193]: Not strange, it's what I'd do. You can see it in `Nop` and `ClientFilesystemWrite` for example.

You need to return a value and it's wasteful to re-fill the "input" value. Tasks only return new values.

File: pkg/dagger.io/dagger/core/exec.cue [hunk 5]
```diff
@@ -125,3 +128,48 @@ import "dagger.io/dagger"
...
+
+// Send a signal to a command created by #Start. SIGKILL is not
+// supported here. Instead, #Stop should be used to forcibly stop
+// a command.
+#SendSignal: {
+	$dagger: task: _name: "SendSignal"
+
+	input: #Start
+
+	signal: or([ for name, signal in _signals {signal}])
```
[the remark this answers is not in the document]
Comment (helderco) [reply pkg/dagger.io/dagger/core/exec.cue:140]: Ah got it. Sorry I didn't notice it wasn't in the list.
